### ISSUE TITLE
Bug: KMS Error Infinite Loop

### ISSUE THREAD

**Issue by kscreen** [ciscospark/spark-js-sdk#658]
We are developing server side APIs using this library.  We need a configurable timeout so that multiple requests spinning forever will not kill our server if the below error occurs.

We saw the following error that repeated over and over again:

kms: request timed out; request id: e4a504e4-54ed-4879-8065-f6f6633f6573; timeout: 6000
kms: request error KmsError: undefined
KMS_REQUEST_ID: e4a504e4-54ed-4879-8065-f6f6633f6573
at KmsError.ExtendableBuiltin (~/node_modules/@ciscospark/common/dist/exception.js:50:44)
at KmsError.Exception (~/node_modules/@ciscospark/common/dist/exception.js:99:142)
at new KmsError (~/node_modules/@ciscospark/internal-plugin-encryption/dist/kms-errors.js:56:123)
at ~/node_modules/@ciscospark/internal-plugin-encryption/dist/kms-batcher.js:163:20
at process._tickCallback (internal/process/next_tick.js:103:7)
mercury: already connected, will not connect again
kms: wrapping ephemeral key request
kms: request payload { body:
{ uri: 'https://encryption-a.wbx2.com/encryption/api/v1/ecdhe',
method: 'create',
jwk:
{ kty: 'EC',
kid: 'gy30QK-srPYAWevaWTa4SYj0k71A-Htn9tNUoW8zduI',
crv: 'P-256',
x: '2och_Qf8PlMguJ0uAuk4JsPYslAA1brEadp9BVFWDDc',
y: 'Kc9FL9XlWmN5mKytZMrBoP1M-bMfV9P7r9EixVzYZ44' },
requestId: 'd21198c4-62a8-476b-ae8b-9a494dd69cef',
client:
{ clientId: 'https://wdm-a.wbx2.com/wdm/api/v1/devices/3d3614d4-ed41-4609-82e2-254d02aedf0e',
credential:
{ userId: '9e545706-56f0-4a07-bebf-b5fad85ef266',
bearer: 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX' } } },
requestId: 'd21198c4-62a8-476b-ae8b-9a494dd69cef',
uri: 'https://encryption-a.wbx2.com/encryption/api/v1/ecdhe',
method: 'create' }
(node:88457) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 9): KmsError: undefined

Eventually, we started seeing this error as well. 

KMS_REQUEST_ID: 308d29eb-7fb9-4d49-9046-7526fb26b297
kms: retrieving KMS cluster
kms: batched-request-length 1
plugin-storage(Device): waiting to init key @
plugin-storage(Device): already inited @
plugin-storage(Device): waiting to init key @
plugin-storage(Device): already inited @
plugin-storage(Device): waiting to init key @
plugin-storage(Device): already inited @
plugin-storage(Device): waiting to init key @
plugin-storage(Device): already inited @
plugin-storage(Credentials): waiting to init key @
plugin-storage(Credentials): already inited @
auth: received 401, attempting to reauthenticate
TypeError: reason.options.body.map is not a function
at child.handleHttpError (~/node_modules/@ciscospark/spark-core/dist/lib/batcher.js:208:54)
at ~/node_modules/@ciscospark/spark-core/dist/lib/batcher.js:142:25
at process._tickCallback (internal/process/next_tick.js:103:7)
(node:35503) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 867): TypeError: reason.options.body.map is not a function
kms: request timed out; request id: e01b1370-958b-4429-8e1b-3111a51a01cd; timeout: 6000
kms: request error KmsError: undefined
KMS_REQUEST_ID: e01b1370-958b-4429-8e1b-3111a51a01cd
at KmsError.ExtendableBuiltin (~/node_modules/@ciscospark/common/dist/exception.js:50:44)
at KmsError.Exception (~/node_modules/@ciscospark/common/dist/exception.js:99:142)
at new KmsError (~/node_modules/@ciscospark/plugin-encryption/dist/kms-errors.js:56:123)
at ~/node_modules/@ciscospark/plugin-encryption/dist/kms-batcher.js:161:20
at process._tickCallback (internal/process/next_tick.js:103:7)

It does not appear that it will ever recover on its own.  
We would like a configurable timeout.

**Comment by ianwremmel** [comment]
The "infinite loop" is by design and should recover on its own. The second error implies there's a bug somewhere preventing the eventual recovery. 

I'd happily accept a PR for a configurable timeout, but it's unlikely that it'll be prioritized such that I can write that code in the near term. That said, I do want to figure out why it didn't eventually recover.

**Comment by ianwremmel** [comment]
#739 should have fixed the `TypeError: reason.options.body.map is not a function` error and #740 should have fixed firing multiple encryption requests in parallel. Are you still seeing this issue?

**Comment by ianwremmel** [comment]
I think pretty much everything in here is either a result of a server limitation or fixed in a PR that'll open later today. Once you see this issue close, please try to latest sdk and open a new issue with steps to reproduce if it happens for anything other than `listAuthorizations()` on a resource with a large number of authorizations. 

See #748 for some more details on `listAuthorizations()`
